### ISSUE TITLE
Cannot run --train on Mac OS

### ISSUE THREAD

**Issue by robertsolero** [Unity-Technologies/ml-agents#2528]
Hello,
I cannot run any --train command on my Mac OSx. 

**Environment**

Mac OS X (Mojave 10.14.5)
Unity 2019.2.0f1
ML-Agents version 0.9.3 [Also replicated with  0.9.2]
Python 3.6.8 [Also replicated with  3.6.1]
Tested using "UnitySDK" Project folder, as well as a custom Environment/Project in Unity. 

**Description**

I successfully installed ML-Agents after some troubles. 
I tested a pre-trained model on Unity and I'm now attempting to train my own model.

When I try to run 
`mlagents-learn ../config/trainer_config.yaml --run-id=firstRun --train`
as well as 
`mlagents-learn ../config/trainer_config.yaml --env=../myExecutable.app --run-id=firstRun --train`

I get the following error(s): 

`socket.gaierror: [Errno 8] nodename nor servname provided, or not known`

`mlagents.envs.exception.UnityWorkerInUseException: Couldn't start socket communication because worker number 0 is still in use. You may need to manually close a previously opened environment or use a different worker number`

`raise EOFError EOFError`

`mlagents.envs.exception.UnityCommunicationException: UnityEnvironment worker: recv failed.`

<details>
  <summary>CLICK TO SHOW FULL ERROR BELOW</summary>
  
UNITY LOGO.

INFO:mlagents.trainers:{'--base-port': '5005',
 '--curriculum': 'None',
 '--debug': False,
 '--docker-target-name': 'None',
 '--env': 'None',
 '--help': False,
 '--keep-checkpoints': '5',
 '--lesson': '0',
 '--load': False,
 '--multi-gpu': False,
 '--no-graphics': False,
 '--num-envs': '1',
 '--num-runs': '1',
 '--run-id': 'firstRun',
 '--sampler': 'None',
 '--save-freq': '50000',
 '--seed': '-1',
 '--slow': False,
 '--train': True,
 '<trainer-config-path>': '/Users/robertsolero/ml-agents/config/trainer_config.yaml'}
Process Process-1:
Traceback (most recent call last):
  File "/Users/robertsolero/ml-agents/ml-agents-envs/mlagents/envs/rpc_communicator.py", line 75, in check_port
    s.bind(("localhost", port))
**socket.gaierror: [Errno 8] nodename nor servname provided, or not known**

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, *self._kwargs)
  File "/Users/robertsolero/ml-agents/ml-agents-envs/mlagents/envs/subprocess_env_manager.py", line 74, in worker
    env = env_factory(worker_id)
  File "/Users/robertsolero/ml-agents/ml-agents/mlagents/trainers/learn.py", line 259, in create_unity_environment
    base_port=start_port,
  File "/Users/robertsolero/ml-agents/ml-agents-envs/mlagents/envs/environment.py", line 80, in __init__
    self.communicator = self.get_communicator(worker_id, base_port, timeout_wait)
  File "/Users/robertsolero/ml-agents/ml-agents-envs/mlagents/envs/environment.py", line 174, in get_communicator
    return RpcCommunicator(worker_id, base_port, timeout_wait)
  File "/Users/robertsolero/ml-agents/ml-agents-envs/mlagents/envs/rpc_communicator.py", line 48, in __init__
    self.create_server()
  File "/Users/robertsolero/ml-agents/ml-agents-envs/mlagents/envs/rpc_communicator.py", line 54, in create_server
    self.check_port(self.port)
  File "/Users/robertsolero/ml-agents/ml-agents-envs/mlagents/envs/rpc_communicator.py", line 77, in check_port
    raise UnityWorkerInUseException(self.worker_id)
**mlagents.envs.exception.UnityWorkerInUseException: Couldn't start socket communication because worker number 0 is still in use. You may need to manually close a previously opened environment or use a different worker number.**
Traceback (most recent call last):
  File "/Users/robertsolero/ml-agents/ml-agents-envs/mlagents/envs/subprocess_env_manager.py", line 55, in recv
    response: EnvironmentResponse = self.conn.recv()
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/connection.py", line 250, in recv
    buf = self._recv_bytes()
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/connection.py", line 407, in _recv_bytes
    buf = self._recv(4)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/connection.py", line 383, in _recv
  **raise EOFError EOFError**

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.6/bin/mlagents-learn", line 11, in <module>
    load_entry_point('mlagents', 'console_scripts', 'mlagents-learn')()
  File "/Users/robertsolero/ml-agents/ml-agents/mlagents/trainers/learn.py", line 337, in main
    run_training(0, run_seed, options, Queue())
  File "/Users/robertsolero/ml-agents/ml-agents/mlagents/trainers/learn.py", line 96, in run_training
    sampler_file_path, env.reset_parameters, run_seed
  File "/Users/robertsolero/ml-agents/ml-agents-envs/mlagents/envs/subprocess_env_manager.py", line 214, in reset_parameters
    return self.env_workers[0].recv().payload
  File "/Users/robertsolero/ml-agents/ml-agents-envs/mlagents/envs/subprocess_env_manager.py", line 58, in recv
    raise UnityCommunicationException("UnityEnvironment worker: recv failed.")
**mlagents.envs.exception.UnityCommunicationException: UnityEnvironment worker: recv failed.**
</details>

After updating python to 3.6.8 the last 2 errors came up, before (Using python 3.6.1) I only got the first 2. 

It looks like it might related to this [ISSUE](https://github.com/Unity-Technologies/ml-agents/issues/1505), the difference is that I cannot run --train even once. Moreover the API and usage seem to be changed substantially. 

I tried to use Jupyter as well and the same issue come up. 
On Jupyter, using the "getting-started" file, I tried to add on the step 3, a parameter worker_id=2 as follow 

`env = UnityEnvironment(file_name=env_name, worker_id=2)

Set the default brain to work with
default_brain = env.brain_names[0]
brain = env.brains[default_brain]`

The console changed the index on the string, but the same error occurred. 

Using the terminal and adding `--worker_id=X` resulted in not having the --train command processed, as it looks like to be removed [HERE](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md#migrating-from-ml-agents-toolkit-v07-to-v08)

Again, using the terminal and adding `--base-port =5040`(or any other number) returns the same error, where the initial console info showed a different base-port.

Not being an expert in python and socket communication, I'm quite stuck and I do not know where to start to close a port, or any other possible fix. 
Hopefully someone can help me out and guide to a resolution. 
Thanks.

**Comment by chriselion** [comment]
Hi @robertsolero,
I think this is a different underlying error than https://github.com/Unity-Technologies/ml-agents/issues/1505 - in that issue, `socket.bind()` raised `OSError: [Errno 98] Address already in use` but yours raises `socket.gaierror: [Errno 8] nodename nor servname provided, or not known`. I think we're incorrectly treating this error with `UnityWorkerInUseException`.

I'm not sure the best way to debug this, but it seems like there are some related stackoverflow questions, e.g.
* https://stackoverflow.com/questions/39970606/gaierror-errno-8-nodename-nor-servname-provided-or-not-known-with-macos-sie (same error, but thrown from different place)
* https://stackoverflow.com/questions/50376329/s-bind-nodename-nor-servname-provided-or-not-known
* https://apple.stackexchange.com/questions/253817/cannot-ping-my-local-machine

Can you try some of the diagnostic steps there and see if you can get any socket binding (outside of mlagents) working first?

**Comment by robertsolero** [comment]
Hi @chriselion, 
Thank you very much again. 
I've been able to solve the issue, which isn't ML-Agent related, but due entirely to my machine. 

To anyone that might face the same issue here what I did. 

Before I couldn't run 
`ping localhost `
but 
`ping 127.0.0.1`
worked as expected. 

I followed these suggestions [HERE](https://apple.stackexchange.com/questions/253817/cannot-ping-my-local-machine) 
and tried the following:

Changing my local hostname, using System Preferences > Sharing > Edit.
`sudo killall -HUP mDNSResponder`
`dscacheutil -flushcache`
`sudo launchctl unload -w /System/Library/LaunchDaemons/com.apple.mDNSResponder.plist`
`sudo launchctl load -w /System/Library/LaunchDaemons/com.apple.mDNSResponder.plist`

Then I changed my hosts file:

`sudo nano /etc/hosts`
Back up my old file contents. 
Removed everything expect
`127.0.0.1 localhost`

Rebooting. 

`ping localhost ` 
Now works as expected.
